### ISSUE TITLE
How to use NTXentLoss as in CPC?

### ISSUE THREAD

**Issue by vgaraujov** [KevinMusgrave/pytorch-metric-learning#179]
Hello! Thanks for this incredible contribution.

I want to know how to use the NTXentLoss as in CPC model. I mean, I have a positive sample and N-1 negative samples.

Thank you for your help in this matter.

**Comment by KevinMusgrave** [comment]
If you have just a single positive pair in your batch:

```python
from pytorch_metric_learning.losses import NTXentLoss
loss_func = NTXentLoss()

# in your training loop
batch_size = data.size(0)
embeddings = your_model(data)
labels = torch.arange(batch_size)
# The assumption here is that data[0] and data[1] are the positive pair
# And there are no other positive pairs in the batch
labels[1] = labels[0]
loss = loss_func(embeddings, labels)
loss.backward()
```

If your batch size is N, and you have N/2 positive pairs:
```python
from pytorch_metric_learning.losses import NTXentLoss
loss_func = NTXentLoss()

# in your training loop
batch_size = data.size(0)
embeddings = your_model(data)
# The assumption here is that data[0] and data[1] are a positive pair
# data[2] and data[3] are the next positive pair, and so on
labels = torch.arange(batch_size)
labels[1::2] = labels[0::2]
loss = loss_func(embeddings, labels)
loss.backward()
```

**Comment by vgaraujov** [comment]
Thank you for your answer @KevinMusgrave 
To be sure how it works...

Regarding this assumption, data[0] and data[1] are positive pairs, so the rest (data[2:]) will be used as negatives samples?

> If you have just a single positive pair in your batch:
> 
> ```python
> from pytorch_metric_learning.losses import NTXentLoss
> loss_func = NTXentLoss()
> 
> # in your training loop
> batch_size = data.size(0)
> embeddings = your_model(data)
> labels = torch.arange(batch_size)
> # The assumption here is that data[0] and data[1] are the positive pair
> # And there are no other positive pairs in the batch
> labels[1] = labels[0]
> loss = loss_func(embeddings, labels)
> loss.backward()
> ```

Does it mean that all examples that are not positive samples in the batch automatically used as negative samples?

**Comment by KevinMusgrave** [comment]
Yes, data[2:] will be used as negative samples, because their labels are different from data[0] and data[1]. And data[0] and data[1] are the only positive pair because there are no other labels that occur more than once.

**Comment by Dreamcreationman** [comment]
I'm still confused that if I have a batch of randomly sampled image and their corresponding label. How can I use NTXentLoss?
`from pytorch_metric_learning.losses import NTXentLoss
loss_func = NTXentLoss()

# in your training loop
batch_size = data.size(0)
embeddings = your_model(data)
labels = torch.arange(batch_size)
# The assumption here is that data[0] and data[1] are the positive pair
# And there are no other positive pairs in the batch
labels[1] = labels[0]
loss = loss_func(embeddings, labels)
loss.backward()
`

From code provided, why can we regard data[0] and data[1] as positive sample while other pairs are negative?

**Comment by Dreamcreationman** [comment]
Then how can NTXentLoss distinguish positive sample or negative? Images are labeled with their own labels.

**Comment by KevinMusgrave** [comment]
Images with the same label form positive pairs, and images with different labels form negative pairs.

For example, if the labels in a batch are [0, 0, 1, 1, 1] then:
- the positive pairs will be formed by indices [0, 1], [1, 0], [2, 3], [2, 4], [3, 2], [3, 4], [4, 2], [4, 3]
- the negative pairs will be formed by indices [0, 2], [0, 3], [0, 4], [1, 2], [1, 3], [1, 4], [2, 0], [2, 1], [3, 0], [3, 1], [4, 0], [4, 1]

**Comment by Dreamcreationman** [comment]
What if there is no same label pairs as batch are sampled randomly?

**Comment by KevinMusgrave** [comment]
Then NTXentLoss will return 0, because it requires positive pairs to compute an actual loss.

**Comment by KevinMusgrave** [comment]
You can try using [MPerClassSampler](https://kevinmusgrave.github.io/pytorch-metric-learning/samplers/#mperclasssampler) to ensure there are positive pairs in every batch.

**Comment by Dreamcreationman** [comment]
Another problem is that when I use MPerClassSampler in my own project, I found that all training data are not shuffled(all labels in one batch are the same). However, shuffle is not allowed when using a sampler.
`
sampler = samplers.MPerClassSampler(data["label_names"], m=4, length_before_new_iter=len(data["image_names"]))

        data_loader_params = dict(sampler = sampler, batch_size = self.batch_size, num_workers = 12, pin_memory = True)
        data_loader = torch.utils.data.DataLoader(dataset, **data_loader_params)
`

Is there anything wrong with my usage?

**Comment by KevinMusgrave** [comment]
What is your batch size? Can you print the batch labels and paste them here?

**Comment by Dreamcreationman** [comment]
Batch Size in data loader is set to 128 and labels pass to MPerClassSampler is as follows:
```
["n01532829","n01558993","n01704323","n01749939","n01770081","n01843383","n01910747","n02074367","n02089867","n02091831",.....,"n07613480"]
```
however , labels in training batch(code: `inputs, targets = inputs.cuda(), targets.cuda()`) with MPerClassSampler are as follows:
```
tensor([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0], device='cuda:0')
```

**Comment by KevinMusgrave** [comment]
The labels should be integers.

Also, in addition to passing the batch size into the dataloader, you can pass ```batch_size``` into MPerClassSampler. Then it will check to make sure your ```m```, ```labels```, and ```batch_size``` are all compatible.

**Comment by KevinMusgrave** [comment]
Make sure that ```labels``` is the same length as your dataset.

**Comment by JiudingYang** [comment]
Hi Kevin, I just found that, if I have a label [0,1,0,2] for [x1,x2,x3,x4], the cos similarity of (x1,x3) will become 1. It seems that I have to sort the list to [0,0,1,2] to make sure the metric works. So I am wondering, will a long label like [0,0,1,1,2,3,4,5,5] works? (Unbalanced classes).

**Comment by ashutoshml** [comment]
> Images with the same label form positive pairs, and images with different labels form negative pairs.
> 
> For example, if the labels in a batch are [0, 0, 1, 1, 1] then:
> 
> * the positive pairs will be formed by indices [0, 1], [1, 0], [2, 3], [2, 4], [3, 2], [3, 4], [4, 2], [4, 3]
> * the negative pairs will be formed by indices [0, 2], [0, 3], [0, 4], [1, 2], [1, 3], [1, 4], [2, 0], [2, 1], [3, 0], [3, 1], [4, 0], [4, 1]

Is there a way to specify negative and positive pairs instead of providing labels? 
i.e.,
```
loss_func(embeddings, positive_pairs, negative_pairs)
```

**Comment by KevinMusgrave** [comment]
Yes, but unfortunately "dummy" labels are still required:

```python
# positive pairs are formed by (a1, p)
# negatives pairs are formed by (a2, n)
a1= torch.randint(0, 10, size=(100,))
p = torch.randint(0, 10, size=(100,))
a2 = torch.randint(0, 10, size=(100,))
n = torch.randint(0, 10, size=(100,))

pairs = a1, p, a2, n
# won't actually be used
labels = torch.zeros(len(embeddings))
loss_func(embeddings, labels, pairs)
```

**Comment by ashutoshml** [comment]
> Yes, but unfortunately "dummy" labels are still required:
> 
> ```python
> # positive pairs are formed by (a1, p)
> # negatives pairs are formed by (a2, n)
> a1= torch.randint(0, 10, size=(100,))
> p = torch.randint(0, 10, size=(100,))
> a2 = torch.randint(0, 10, size=(100,))
> n = torch.randint(0, 10, size=(100,))
> 
> pairs = a1, p, a2, n
> # won't actually be used
> labels = torch.zeros(len(embeddings))
> loss_func(embeddings, labels, pairs)
> ```

Thanks.

"the positive pairs will be formed by indices [0, 1], [1, 0], [2, 3], [2, 4], [3, 2], [3, 4], [4, 2], [4, 3]
the negative pairs will be formed by indices [0, 2], [0, 3], [0, 4], [1, 2], [1, 3], [1, 4], [2, 0], [2, 1], [3, 0], [3, 1], [4, 0], [4, 1]"

So the number of pairs (a1, p) can be different from (a2, n) as in this example, right? As in 

```
a1= torch.randint(0, 10, size=(100,))
p = torch.randint(0, 10, size=(100,))
a2 = torch.randint(0, 10, size=(500,))
n = torch.randint(0, 10, size=(500,))
```

**Comment by KevinMusgrave** [comment]
Yes

**Comment by lucasestini** [comment]
Hello, does the loss consider _every_ positive pair in the batch? Like, if I have 3 samples belonging to the same class, do they all contribute to the loss and get pulled together? Are they considered as 6 positives pairs or treated at once?

**Comment by KevinMusgrave** [comment]
@lucasestini They are considered as 6 positive pairs.

**Comment by YK711** [comment]
Hi @KevinMusgrave , thank you for your wonderful code. I found that it still requires dummy labels as input. ```pytorch_metric_learning``` version is 1.6.2. PyTorch version is 1.12.0. I use DDP. Thanks!

**Comment by KevinMusgrave** [comment]
@YK711 Are you using `DistributedLossWrapper`? I see I forgot to update that class to make `labels` optional. I've added it to my todo list now: https://github.com/KevinMusgrave/pytorch-metric-learning/issues/531

**Comment by YK711** [comment]
Yes, I use ```DistributedLossWrapper```. Thank you @KevinMusgrave!

**Comment by KevinMusgrave** [comment]
`SelfSupervisedLoss` has been added to v2.0.0. If you have two "views" of data, you don't need to make labels anymore:

```python
from pytorch_metric_learning.losses import SelfSupervisedLoss
loss_func = SelfSupervisedLoss(NTXentLoss())
embeddings = model(data)
augmented = model(augmented_data)
loss = loss_func(embeddings, augmented)
```

**Comment by happen2me** [comment]
I find the formula in the documentation for the [NTXentLoss](https://kevinmusgrave.github.io/pytorch-metric-learning/losses/#ntxentloss) is misleading:
![image](https://user-images.githubusercontent.com/11248271/230785076-089479db-a81b-496b-881a-f2bb7698eaf9.png)


If labels for indices `[a, b, c, d]` are `[0, 0, 2, 3]`, the formula implies that the positive pair is `[a, b]`, while the negative pairs are `[a, c], [a, d]`. However, this is different from its actual implementation, where the positive pairs are `[a, b], [b, a]`, and the negative samples are `[a, c], [a, d], [b, c], [b, d], [c, a], [c, b], [c, d], [d, a], [d, b], [d, c]`.

Maybe we should remove this formula and replace it with more clear sampling information? I can make a pull request for it if I am not wrong :)

**Comment by KevinMusgrave** [comment]
> If labels for indices [a, b, c, d] are [0, 0, 2, 3], the formula implies that the positive pair is [a, b], while the negative pairs are [a, c], [a, d].

@happen2me Actually that _is_ how it works. 
The negative pairs for `[a, b]` will all be of the form `[a, _]`. 
The negative pairs for `[b, a]` will all be of the form `[b, _]`.

I apologize if my comments further up this thread were confusing.

See these related comments:

https://github.com/KevinMusgrave/pytorch-metric-learning/discussions/606#discussioncomment-5542877
https://github.com/KevinMusgrave/pytorch-metric-learning/issues/600#issuecomment-1483894037
https://github.com/KevinMusgrave/pytorch-metric-learning/issues/6#issuecomment-587331886

**Comment by happen2me** [comment]
@KevinMusgrave Thank you very much for your reply! Yes, when I saw the equation, I thought the positive pairs are `[anchor, positive]`, and the negative pairs are `[anchor, negative 1]`, `[anchor, negative 2]`... But in fact, those of the same label with `anchor` are regarded positive samples to each other (including the anchor), while the negative samples also include pairs like `[positive, negative k]`, and `[negative k1, negative k2]`, which is not that intuitive purely from the equation.

I wanted to apply the loss in a situation like `[question, positive passage, negative passage 1, negative passage 2, ...]`. In this case, pushing negatives apart from each other isn't really necessary. I managed to do it with manually created `indices_tuple`.

**Comment by KevinMusgrave** [comment]
> I wanted to apply the loss in a situation like `[question, positive passage, negative passage 1, negative passage 2, ...]`. In this case, pushing negatives apart from each other isn't really necessary. I managed to do it with manually created `indices_tuple`.

Just so we're on the same page, I'll go through a simple example.

Say we have a batch size of 4 with labels: `[dog, dog, cat, mouse]`.

In this case there will be two losses:

- one loss for the positive pair `[0, 1]` with negative pair indices `[0, 2], [0,3]`.
- one loss for the positive pair `[1, 0]` with negative pair indicies `[1, 2], [1,3]`.

So the `[2,3]` (cat, mouse) isn't used at all.

I've created an issue for improving the documentation: https://github.com/KevinMusgrave/pytorch-metric-learning/issues/608
